src/psyclone/tests/domain/lfric/lfric_scalar_stubgen_test.py

```diff
@@ -0,0 +1,128 @@
+LFRic scalar arguments.
+'''
+
+from __future__ import absolute_import, print_function
+import os
+import pytest
+
+from fparser import api as fpapi
+from psyclone.domain.lfric import LFRicConstants
+from psyclone.dynamo0p3 import DynKernMetadata, DynKern, LFRicScalarArgs
+from psyclone.f2pygen import ModuleGen
+from psyclone.errors import InternalError
+from psyclone.gen_kernel_stub import generate
+from psyclone.parse.utils import ParseError
+
+# Constants
+BASE_PATH = os.path.join(
+    os.path.dirname(os.path.dirname(os.path.dirname(
+        os.path.abspath(__file__)))),
+    "test_files", "dynamo0p3")
+TEST_API = "dynamo0.3"
+
+
+def test_lfricscalars_stub_err():
+    ''' Check that LFRicScalarArgs._stub_declarations() raises the
+    expected internal error if it encounters an unrecognised data
+    type of a scalar argument when generating a kernel stub.
+
+    '''
+    ast = fpapi.parse(os.path.join(BASE_PATH,
+                                   "testkern_one_int_scalar_mod.f90"),
+                      ignore_comments=False)
+    metadata = DynKernMetadata(ast)
+    kernel = DynKern()
+    kernel.load_meta(metadata)
+    # Sabotage the scalar argument to make it have an invalid data type
+    arg = kernel.arguments.args[1]
+    arg.descriptor._data_type = "gh_invalid_scalar"
+    with pytest.raises(InternalError) as err:
+        LFRicScalarArgs(kernel)._stub_declarations(ModuleGen(name="my_mod"))
+    const = LFRicConstants()
+    assert ("Found an unsupported data type 'gh_invalid_scalar' for the "
+            "scalar argument 'iscalar_2'. Supported types are {0}.".
+            format(const.VALID_SCALAR_DATA_TYPES)
+            in str(err.value))
+
+
+SIMPLE_WITH_SCALARS = (
+    "  MODULE simple_with_scalars_mod\n"
+    "    IMPLICIT NONE\n"
+    "    CONTAINS\n"
+    "    SUBROUTINE simple_with_scalars_code(nlayers, rscalar_1, field_2_w1, "
+    "iscalar_3, lscalar_4, ndf_w1, undf_w1, map_w1)\n"
+    "      USE constants_mod, ONLY: r_def, l_def, i_def\n"
+    "      IMPLICIT NONE\n"
+    "      INTEGER(KIND=i_def), intent(in) :: nlayers\n"
+    "      INTEGER(KIND=i_def), intent(in) :: ndf_w1\n"
+    "      INTEGER(KIND=i_def), intent(in), dimension(ndf_w1) :: map_w1\n"
+    "      INTEGER(KIND=i_def), intent(in) :: undf_w1\n"
+    "      REAL(KIND=r_def), intent(in) :: rscalar_1\n"
+    "      INTEGER(KIND=i_def), intent(in) :: iscalar_3\n"
+    "      LOGICAL(KIND=l_def), intent(in) :: lscalar_4\n"
+    "      REAL(KIND=r_def), intent(inout), dimension(undf_w1) ::"
+    " field_2_w1\n"
+    "    END SUBROUTINE simple_with_scalars_code\n"
+    "  END MODULE simple_with_scalars_mod")
+
+
+def test_stub_generate_with_scalars():
+    ''' check that the stub generate produces the expected output when
+    the kernel has scalar arguments '''
+    result = generate(os.path.join(BASE_PATH, "simple_with_scalars.f90"),
+                      api=TEST_API)
+    print(str(result))
+    assert SIMPLE_WITH_SCALARS in str(result)
```

Comment:
As SIMPLE_WITH_SCALARS is only used in this test, please could you move it into here.

Reply:
Done (now with the "testkern_three_scalars_mod.f90".